apps/system/js/data_connection_manager.js

```diff
@@ -0,0 +1,79 @@
+/* global BaseModule, LazyLoader, RoamingWarningSystemDialog */
+'use strict';
+
+(function() {
+  // Responsible to load and init the sub system for mobile connections.
+  var DataConnectionManager = function(core) {
+    this.core = core;
+    this.mobileConnections = core.mobileConnections;
+  };
+
+  DataConnectionManager.SETTINGS = [
+    'ril.data.roaming_enabled',
+    'ril.data.roaming_enabled.warningDialog.enabled'
+  ];
+
+  DataConnectionManager.SERVICES = [
+    'enableDataConnection'
+  ];
+
+  BaseModule.create(DataConnectionManager, {
+    name: 'DataConnectionManager',
+    DEBUG: false,
+
+    checkRoaming: function(conn) {
+      var roaming = this._settings['ril.data.roaming_enabled'];
+      return this.DEBUG || (!roaming && conn.data.roaming);
+    },
+
+    enableDataConnection: function() {
+      var dataType, conn;
+      for (var i = 0; i < this.mobileConnections.length && !dataType; i++) {
+        dataType = this.mobileConnections[i].data.type;
+        conn = this.mobileConnections[i];
+      }
+      if (!dataType) {
+        // No connection available
+        this.debug('No connection available', dataType);
+        return;
+      }
+
+      if (this.checkRoaming(conn)) {
+        this.disabledDefaultDialogIfNeeded();
```

Comment:
It seems weird that we call a function called `disabledDefaultDialogIfNeeded` that both disables the dialog (if needed) but then also displays dialog. I think it would better here we called something like `showRoamingPromptIfNeeded`, and then that function called `disabledDefaultDialogIfNeeded` (which just deals with the setting) and then checks the setting and displays the dialog if needed. Right now we don't even check the setting, so we are always displaying this dialog.